be/src/exec/vectorized/es_http_components.cpp

```diff
@@ -0,0 +1,540 @@
+            //                    "_id": "UhHNc3IB8XwmcbhBk1ES",
+            //                    "_source": {
+            //                          "k": 201,
+            //                    }
+            //                }
+            //            ]
+            //        }
+            if (slot_desc->col_name() == FIELD_ID) {
+                // actually this branch will not be reached, this is guaranteed by Doris FE.
+                if (pure_doc_value) {
+                    return Status::RuntimeError("obtain `_id` is not supported in doc_values mode");
+                }
+                PrimitiveType type = slot_desc->type().type;
+                DCHECK(type == TYPE_CHAR || type == TYPE_VARCHAR);
+
+                const auto& _id = obj[FIELD_ID];
+                Slice slice(_id.GetString(), _id.GetStringLength());
+                _append_data<TYPE_VARCHAR>(column.get(), slice);
+
+                continue;
+            }
+
+            // if pure_doc_value enabled, docvalue_context must contains the key
+            // todo: need move all `pure_docvalue` for every tuple outside fill_tuple
+            //  should check pure_docvalue for one table scan not every tuple
+            const char* col_name = pure_doc_value ? _docvalue_context->at(slot_desc->col_name()).c_str()
+                                                  : slot_desc->col_name().c_str();
+
+            auto has_col = line.HasMember(col_name);
+            if (has_col) {
+                const rapidjson::Value& col = line[col_name];
+                // doc value
+                bool is_null = (pure_doc_value && col.IsArray() && col[0].IsNull()) || col.IsNull();
+                if (!is_null) {
+                    // append value from ES to column
+                    _append_value_from_json_val(column.get(), slot_desc->type().type, col, pure_doc_value);
+                    continue;
+                }
+                // handle null col
+                if (slot_desc->is_nullable()) {
+                    _append_null(column.get());
+                } else {
+                    return Status::DataQualityError(
+                            fmt::format("col `{}` is not null, but value from ES is null", slot_desc->col_name()));
+                }
+            } else {
+                // if don't has col in ES , append a default value
+                _append_null(column.get());
+            }
+        }
+    }
+    _cur_line += fill_sz;
+    return Status::OK();
+}
+
+void ScrollParser::set_params(const TupleDescriptor* descs,
+                              const std::map<std::string, std::string>* docvalue_context) {
+    _tuple_desc = descs;
+    _docvalue_context = docvalue_context;
+}
+
+bool ScrollParser::_pure_doc_value(const rapidjson::Value& obj) {
+    if (obj.HasMember("fields")) {
+        return true;
+    }
+    return false;
+}
+
+template <PrimitiveType type, typename CppType>
+void ScrollParser::_append_data(Column* column, CppType& value) {
```

Comment:
Json related implementations can be found at be/src/exec/vectorized/json_scanner.h if it can be reused

Reply:
It is difficult to reuse logic in the json scanner